xwiki-rendering-syntaxes/xwiki-rendering-syntax-xwiki21/src/main/java/org/xwiki/rendering/internal/listener/TmpParseListenerProvider.java

```diff
@@ -0,0 +1,144 @@
+
+    private static final List<String> ACCEPTED_SYNTAX = List.of(
+        XWIKI_2_0.toIdString(),
+        XWIKI_2_1.toIdString()
+    );
+
+    private static class InternalChainingListener extends AbstractChainingListener
+    {
+        private static final List<String> KNOWN_PARAMETERS = List.of(
+            WIDTH_PROPERTY,
+            STYLE_PROPERTY,
+            // TODO: reuse constant if it exists
+            "data-xwiki-image-style",
+            "data-xwiki-image-style-alignment",
+            "data-xwiki-image-style-border",
+            "data-xwiki-image-style-text-wrap"
+        );
+
+        private Map<String, String> additionalFigureParameters;
+//
+
+        /**
+         * TODO.
+         *
+         * @param listenerChain TODO
+         */
+        protected InternalChainingListener(ListenerChain listenerChain)
+        {
+            setListenerChain(listenerChain);
+        }
+
+        @Override
+        public void onImage(ResourceReference reference, boolean freestanding, String id,
+            Map<String, String> parameters)
+        {
+            this.additionalFigureParameters = new HashMap<>();
+            KNOWN_PARAMETERS.forEach(knowParameter -> {
+                if (parameters.containsKey(knowParameter)) {
+                    String param = parameters.get(knowParameter);
+                    if (Objects.equals(knowParameter, WIDTH_PROPERTY)) {
+                        this.additionalFigureParameters.put(STYLE_PROPERTY, String.format("width: %spx;", param));
+                    } else {
+                        this.additionalFigureParameters.put(knowParameter, param);
+                    }
+                }
+            });
+
+            super.onImage(reference, freestanding, id, parameters);
+        }
+
+        @Override
+        public void endFigure(Map<String, String> parameters)
+        {
+            Map<String, String> withImageParams = new HashMap<>(parameters);
+            String existingStyle = withImageParams.get(STYLE_PROPERTY);
+            if (this.additionalFigureParameters != null) {
+                this.additionalFigureParameters.forEach((key, value) -> {
+                    String computedValue;
+                    if (key.equals(STYLE_PROPERTY) && existingStyle != null) {
+                        String format;
+                        if (existingStyle.endsWith(";")) {
+                            format = "%s %s";
+                        } else {
+                            format = "%s; %s";
+                        }
+                        computedValue = String.format(format, existingStyle, value);
+                    } else {
+                        computedValue = value;
+                    }
+                    withImageParams.put(key, computedValue);
+                });
+                withImageParams.putAll(this.additionalFigureParameters);
+                this.additionalFigureParameters = null;
+            }
+            super.endFigure(withImageParams);
```

Comment:
You need to pass these same arguments also to `beginFigure`. I think what you need to do is to start queuing when you encounter `beginFigure` and only forward the events when you've reached `endFigure`. You can achieve this by inheriting from `LookaheadChainingListener` which supports queuing, see `StackingFigureContentChainingListener` for an example.

Reply:
For example, the HTML renderer only prints the arguments that are passed to `beginFigure`, the arguments to `endFigure` are ignored.